### ISSUE TITLE
Make all tests pass on an installed version of Twisted on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7838]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7838|
|Type|enhancement|
|Created|2015-03-27 04:00:36Z|
|Branch|https://github.com/twisted/twisted/tree/py3tests-venv-7838|

This also includes a modified python3 tests script, so that this is easily testable.

<details><summary>Searchable metadata</summary>

```
trac-id__7838 7838
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_py3tests_venv_7838 branches/py3tests-venv-7838
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1427428836233993 1427428836233993
changetime__1429450419461763 1429450419461763
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44294]) Branching to py3tests-venv-7838.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

itamarst thought this was a good idea, so here it is.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_4"></a>|@derwolfe commented|
|-|-|

Hawkowl - thanks for working on this!

1. Could explain why the tests need to run in a virtualenv?
2. If the tests do require the usage of a virtualenv, shouldn't Twisted have a
   hard dependency on virtualenv? Or is this not required since pyvenv ships
   with py3?
3. It seems like there are several tickets here:
   a. use a virtualenv to run py3 tests
   b. move the MemoryReactor and MemoryDatagramTransport from twisted/names/test/test_rootresolve.py to twisted/names/test/test_util.py
   c. add py3 imports to twisted/trial/_asyncrunner.py
   d. add deprecation warning to twisted/test/twisted.py regarding lore

However, if you think I'm not seeing the bigger picture/goal of the ticket, please tell me.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)|@derwolfe set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

Thanks for the review!

- The tests run in the virtualenv to make sure that the tests pass and Twisted works when installed. If a test imports an unported module, it means that it may pass without it being in a venv, and will NOT work when it is actually released.
- This is an admin thing -- for porters and developers. It is not required by anyone else, so it does not need to be a dependency.
- I will break the code changes up into other tickets when I get a chance.

Thanks herrwolfe! :)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

1. If we rename this ticket "make sure python 3 support actually installs all modules correctly", items A, B and C could stay in this issue. D does seem separate, yes.
2. The important thing is that *buildbot* run tests via installation. So either there needs to be a corresponding change to buildbot, or maybe new script should have name of current one, if that's what buildbot is currently running.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

D is a separate issue, yeah. And yes, I'll rename the ticket -- the modified scripts are to ensure that.

When this lands, I'm planning to put a PR for changing the Buildbot to run this instead of the usual one. The non-in-venv one is handy for quick testing during porting, with the venv one to make sure that it works in an installation (as it takes a while to reinstall). I think this sounds reasonable?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl removed owner|
|-|-|

I think D) belongs in this ticket, since this ticket really is a) making sure that Twisted's tests pass on an installed version of Twisted, and b) providing a script to test that.

Putting up for review again.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes looks good and I am ok with including D in this ticket.

py3 tests fail

```
Traceback (most recent call last):
  File "admin/run-python3-tests", line 36, in <module>
    path = os.path.join(twistedPath, "twisted", "python", "dist3.py")
NameError: name 'twistedPath' is not defined
```

--------

I am ok with the new admin script but for the future, rather than multiple script in admin, each called by buildbot I am thinking that we should have a buildbot_dispatcher script which is supposed to execute "standard" buildbot steps like print deps versions, updated deps/create venv or run tests

In this case we can have standard steps for all builders and the buildbot_dispatcher will execute different tests based on environment variables updated by each builder.

thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl removed owner|
|-|-|

Hmm, I seemed to have half-refactored the two different files. Fixed, and now it should work properly.

Respinning builders.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

Changes looks good. +1 to merge them.

I am ok with merging D with this patch.

I am leaving this for review so that itamar or herrwolfe can cast their vote.

I am working on a builder which is executed in a virtualenv and with that I hope that we can put this code under tests.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_14"></a>|@hynek commented|
|-|-|

- I may or may not have given some completely superficial comments that have been addressed in r44499.
- the errors in https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2801/steps/run-twistedchecker/logs/new%20twistedchecker%20errors are benign b/c the affected classes implement interfaces.
- debian6-x86_64-py2.6 and ubuntu-12.04-x86_64-py2.7 fail because our threading tests are broken and unrelated.
- the rest of the bots are green

----

please merge.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)|@hynek set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44500]) Merge py3tests-venv-7838: Ensure that Twisted installs on Python 3 correctly

Author: hawkowl
Reviewers: herrwolfe, adiroiban, hynek
Fixes: [#7838](https://github.com/twisted/twisted/issues/7838)
